File: backend/pkg/database/models/page.go
package models

import (
	"encoding/json"
	"time"

	"backend/pkg/core/page"

	"github.com/google/uuid"
	"gorm.io/datatypes"
)

// Page is the GORM model for pages
type Page struct {
	ID          uuid.UUID      `json:"id" gorm:"type:uuid;primaryKey;default:uuid_generate_v4()"`
	Slug        string         `json:"slug" gorm:"not null;uniqueIndex"`
	Title       string         `json:"title" gorm:"not null"`
	Content     string         `json:"content" gorm:"type:text"`
	Description string         `json:"description"`
	ImageURL    string         `json:"image_url"`
	MetaData    datatypes.JSON `json:"meta_data" gorm:"type:jsonb;default:'{}'"`
	IsPublished bool           `json:"is_published" gorm:"default:true"`
	CreatedAt   time.Time      `json:"created_at" gorm:"autoCreateTime"`
	UpdatedAt   time.Time      `json:"updated_at" gorm:"autoUpdateTime"`
}

func (Page) TableName() string {
	return "page"
}

// ToCore converts the GORM model to the domain type
func (m *Page) ToCore() *page.Page {
	var metaData map[string]interface{}
	if m.MetaData != nil {
		_ = json.Unmarshal(m.MetaData, &metaData)
	}

	return &page.Page{
		ID:          m.ID,
		Slug:        m.Slug,
		Title:       m.Title,
		Content:     m.Content,
		Description: m.Description,
		ImageURL:    m.ImageURL,
		MetaData:    metaData,
		IsPublished: m.IsPublished,
		CreatedAt:   m.CreatedAt,
		UpdatedAt:   m.UpdatedAt,
	}
}

// PageFromCore creates a GORM model from the domain type
func PageFromCore(p *page.Page) *Page {
	var metaData datatypes.JSON
	if p.MetaData != nil {
		metaData, _ = datatypes.NewJSONType(p.MetaData).MarshalJSON()
	}

	return &Page{
		ID:          p.ID,
		Slug:        p.Slug,
		Title:       p.Title,
		Content:     p.Content,
		Description: p.Description,
		ImageURL:    p.ImageURL,
		MetaData:    metaData,
		IsPublished: p.IsPublished,
		CreatedAt:   p.CreatedAt,
		UpdatedAt:   p.UpdatedAt,
	}
}

File: backend/src/database/models/page.rs
use chrono::{DateTime, Utc};
use diesel::{Identifiable, Insertable, Queryable, Selectable};
use serde_json::Value;
use uuid::Uuid;

use crate::{core::page::Page, error::AppError, schema::page};

#[derive(Debug, Clone, Queryable, Selectable, Identifiable)]
#[diesel(table_name = page)]
#[diesel(check_for_backend(diesel::pg::Pg))]
pub struct PageRow {
    pub id: Uuid,
    pub slug: String,
    pub title: String,
    pub content: Option<String>,
    pub description: Option<String>,
    pub image_url: Option<String>,
    pub meta_data: Option<Value>,
    pub is_published: Option<bool>,
    pub created_at: Option<DateTime<Utc>>,
    pub updated_at: Option<DateTime<Utc>>,
}

#[derive(Debug, Insertable)]
#[diesel(table_name = page)]
pub struct NewPageRow<'a> {
    pub id: Uuid,
    pub slug: &'a str,
    pub title: &'a str,
    pub content: &'a str,
    pub description: &'a str,
    pub image_url: &'a str,
    pub meta_data: Option<Value>,
    pub is_published: Option<bool>,
}

impl TryFrom<PageRow> for Page {
    type Error = AppError;

    fn try_from(row: PageRow) -> Result<Self, Self::Error> {
        let meta_data = match row.meta_data {
            Some(Value::Object(values)) => Some(values.into_iter().collect()),
            Some(_) => None,
            None => None,
        };
        Ok(Self {
            id: row.id,
            slug: row.slug,
            title: row.title,
            content: row.content.unwrap_or_default(),
            description: row.description.unwrap_or_default(),
            image_url: row.image_url.unwrap_or_default(),
            meta_data,
            is_published: row.is_published.unwrap_or(false),
            created_at: row.created_at,
            updated_at: row.updated_at,
        })
    }
}
